final-project/application.py

```diff
@@ -563,28 +565,139 @@ def parse_posted_json():
+    if 'reference-form' in requested_product_updates:
+
+        print("references found in product updates")
+
+        # TODO store the reference-form items
+
+
+        # BUG: The following code loops through all the edits returned, not only the reference items.
+        # I thinkkk this means that in certain circumstances v is a string.
+
+        # Loop through each reference
+        for k, v in requested_product_updates.items():
+
+            if (type(v) == dict) and (k == 'reference-form'):
+
+                print('')
+                print('')
+                print('Reference edits')
+                print (v, ' type:', type(v))
+                print('')
+
+                # Loop through the titles and links for each reference.
+                for key, value in v.items():
+
+                    # Look for the reference's id.
+                    if (key == 'referenceId'):
+
+                        print('')
+                        print('Key = ', key)
+                        print('Value =', value)
+                        print('')
+
+                        # Grab the reference id
+                        look_for_reference_id = re.search('\d+', value)
+
+                        reference_id = look_for_reference_id.group()
+                        print('')
+                        print('Parsing only the reference id...')
+                        print(reference_id)
+                        print('')
+
+                    # Check for "deleteReference" value
+                    if (value == 'deleteReference') and (key == 'referenceTitle'):
+
+                        print('')
+                        print('About to delete reference')
+                        print('')
+
+                        # Delete the reference with a matching ref-id and product-id
+                        number_of_updates = db.execute("""
+                            DELETE FROM research
+                            WHERE research.product_id=:product_id
+                            AND research.id=:reference_id"""
+                            , product_id=requested_product_updates['product-id']
+                            , reference_id=reference_id)
+
+                        if (number_of_updates > 0):
+                            print('The reference was successfully deleted')
+                        else:
+                            print('The reference was not deleted')
+
+                    # If the value is 'deleteReference' and we are on the link iteration,
+                    # then this reference data has already been deleted.
+                    elif (value == 'deleteReference') and (key == 'referenceLink'):
+
+                        # Skip this iteration.
+                        continue
+
+                    # Check if an update was made to the reference title.
+                    elif (key == 'referenceTitle'):
+
+                        print('')
+                        print("Updating reference title")
```

Comment:
I notice you like to do these prints. You can also add `\n` into your string to have line breaks. You can also have debug lines that you can pass in as an argument to your python script, so that you only `print` when debug mode is on.

Reply:
I'll definitely start to use \n and look into the argument